retrieval/k8s/k8s_stats.go

```diff
@@ -1,82 +1,102 @@
 package k8s
 
 import (
-	"net/http"
+	"bufio"
+	"encoding/json"
+	"errors"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
 	"os"
-
-	"github.com/cloudability/metrics-agent/retrieval/raw"
-	log "github.com/sirupsen/logrus"
+	"time"
 )
 
-//GetK8sMetrics returns cloudabilty measurements retrieved from a given K8S Clientset
-func GetK8sMetrics(clusterHostURL string, clusterVersion float64, workDir *os.File, rawClient raw.Client) (err error) {
-
-	v1Sources, v1beta1Sources, v1AppSources := getk8sSourcePaths(clusterVersion)
+func StartUpInformers(clientset kubernetes.Interface, clusterVersion float64,
+	resyncInterval int) (map[string]*cache.SharedIndexInformer, error) {
+	factory := informers.NewSharedInformerFactory(clientset, time.Duration(resyncInterval)*time.Hour)
 
-	// get v1 sources
-	for _, v1s := range v1Sources {
-		_, err := rawClient.GetRawEndPoint(http.MethodGet, v1s, workDir, clusterHostURL+"/api/v1/"+v1s, nil, true)
-		if err != nil {
-			log.Errorf("Error retrieving "+v1s+" metric endpoint: %s", err)
-			return err
-		}
+	// v1Sources
+	replicationControllerInformer := factory.Core().V1().ReplicationControllers().Informer()
+	servicesInformer := factory.Core().V1().Services().Informer()
+	nodesInformer := factory.Core().V1().Nodes().Informer()
+	podsInformer := factory.Core().V1().Pods().Informer()
+	persistentVolumesInformer := factory.Core().V1().PersistentVolumes().Informer()
+	persistentVolumeClaimsInformer := factory.Core().V1().PersistentVolumeClaims().Informer()
+	namespacesInformer := factory.Core().V1().Namespaces().Informer()
+	// AppSources
+	replicasetsInformer := factory.Apps().V1().ReplicaSets().Informer()
+	daemonsetsInformer := factory.Apps().V1().DaemonSets().Informer()
+	deploymentsInformer := factory.Apps().V1().Deployments().Informer()
+	// Jobs
+	jobsInformer := factory.Batch().V1().Jobs().Informer()
+	// Cronjobs were introduced in k8s 1.21 so for older versions do not attempt to create an informer
+	var cronJobsInformer cache.SharedIndexInformer
+	if clusterVersion > 1.20 {
+		cronJobsInformer = factory.Batch().V1().CronJobs().Informer()
 	}
+	// closing this will kill all informers
```

Comment:
Ideally we should be closing this channel as well in order to allow the application/informer system and all related goroutines to cleanly shut down. If this could be either passed in from the main method or returned from here to a `defer close(stopCh)` that might be a bit safer.

Reply:
I bumped the close to the CollectKubeMetrics function which should only ever complete in the case of the metrics-agent shutting down. let me know if this was incorrect implementation.